Surround Rewrite

sql-koala opened the pull request:
**What this PR does / why we need it**:
Here is something cool: a complete rewrite of the surround plugin.

Features:
- decoupled from yank operator (possible with new remapper)
- "mini state machine" on SurroundInputMode drastically reduced
- integrates much better with modehandler, so that
  - requires no special logic for repeat
  - supports multiple cursors (with no overhead)
- all exec methods fit on less than one screen; made of 3-4 easy to understand steps because
- handling of edge cases and ugly details are extracted and reused in a few methods
- there is exactly one let (vs const) in the whole code

**Which issue(s) this PR fixes**
fixes #1818
fixes #5211 
fixes #2893
fixes #6564 
should fix #3386

**Special notes for your reviewer**:
This is ready to merge from my side. 
...
And i had to disable the tests with tags, because this uses an inputBox now, to read the tag name.

Looking forward to your feedback @J-Fields ;-)

sql-koala commented:
One remark one visual surround: we have to balance/choose between #5081 vs #5211.

setting this on Visual Operators
`cursorStop = start`
reduces the selection to one line.

Therefore I prefer
`cursorStop = end`

J-Fields commented on test/plugins/surround.test.ts:
```diff
@@ -59,6 +77,7 @@ suite('surround plugin', () => {
     end: ['first |( line ).test'],
   });
 
+  /*
```
Maybe we can stub `vscode.window.showInputBox` to make these tests possible?

sql-koala replied:
I have adapted the existing tests with a stub.

J-Fields commented on src/state/recordedState.ts:
```diff
@@ -184,10 +184,6 @@ export class RecordedState {
 
   public hasRunOperator = false;
 
-  public hasRunSurround = false;
```
Don't have time yet for a thorough review, but I love seeing jank like this being ripped out!

sql-koala commented:
I have added two functionality at other places.
- to make it ship-able: a way to provide default loaded mappings for plugins
- to make it testable: support stub in tests. tag tests helped, I fixed a small bug and a not covered feature.
 
This builds and passes locally.

J-Fields commented:
Sweet! Can you fix the merge conflicts when you get a chance?

sql-koala commented:
@J-Fields I had to wrestle quiet a bit with git, but that should to it.

J-Fields commented:
> One remark one visual surround: we have to balance/choose between #5081 vs #5211.
> 
> setting this on Visual Operators
> `cursorStop = start`
> reduces the selection to one line.
> 
> Therefore I prefer
> `cursorStop = end`

That's fine, at least for now. I think we can eventually hack it so they're both as expected, but don't worry about that for this PR.

sql-koala commented:
Hi @J-Fields, can we get this reviewed and merged?
I just added a few more tests in the last commits, but it is ready from my side.

Because it's a complete rewrite, even small automated changes like
https://github.com/VSCodeVim/Vim/commit/3f6ba739445d469f43324b2757096f3c15c43a5a#diff-9102ca6b9529b6f1b3372b511a9687cd3a0b3cb0229a2cbc4b4c01577cced66d
can result in weird merge conflicts ... which i than need to manage.

J-Fields commented on test/plugins/surround.test.ts:
```diff
@@ -32,6 +33,20 @@ suite('surround plugin', () => {
     end: ['first |(line) test'],
   });
 
+  newTest({
```
`ysw` doesn't work in some cases. For instance, put the cursor at the start of `newTest` here and do `ysw"`

J-Fields replied:
Seems like this happens if there's a newline or punctuation before whitespace

sql-koala replied:
See https://github.com/VSCodeVim/Vim/pull/6452/commits/20629ff7fd36893454179ce27d737b278bfbd7c6 with whole function as context.
Apparently, we are expected to correct w/aw motions, which is a bit ugly.
I thought about/would prefer adding maps: ysw -> yse but that would not work with counts: ys4w.

This seems to handle all known cases correct.

J-Fields commented on src/actions/plugins/pluginDefaultMappings.ts:
```diff
@@ -0,0 +1,38 @@
+    {
+      mode: 'normalModeKeyBindingsNonRecursive',
+      configSwitch: 'surround',
+      mapping: { before: ['y', 's'], after: ['<plugys>'] },
```
Leaving this as-is for now, planning to come back to it at some point (but will likely forget, if I'm being honest)

J-Fields commented:
Thanks a ton for this, @sql-koala! To be honest I just skimmed the code, as this is an area I'm very unfamiliar with, but it seems reasonable, works better than before from what I can tell, and leaks fewer implementation details into ModeHandler and such - good enough for me.
